packages/alfa-rules/src/common/predicate/is-at-the-start.ts

```diff
@@ -0,0 +1,130 @@
+ * only perceivable actual content is bad.
+ *
+ * Note that the def is actually symmetrical, but N2 is conceptually a
+ * container (`<main>`) while N1 is a single point (an anchor),
+ * hence the asymmetry in the code.
+ *
+ * Complexity: the size of the subtree anchored at the lowest common ancestor.
+ */
+export function isAtTheStart(
+  node2: Node,
+  device: Device = Device.standard()
+): Predicate<Node> {
+  const options = { flattened: true, nested: true };
+  const isPerceivableContent = and(isPerceivable(device), isContent(options));
+
+  return function isAtTheStart(node1: Node): boolean {
+    if (node2.equals(node1)) {
+      return true;
+    }
+
+    const context = lowestCommonAncestor(node1, node2, options);
+
+    if (context.isNone()) {
+      // the nodes are not even in the same tree…
+      return false;
+    }
+
+    // Get descendants of the LCA, and skip everything before both nodes.
+    let descendants = context
+      .get()
+      .inclusiveDescendants(options)
+      .skipUntil(or(equals(node1), equals(node2)));
+
+    // node1 and node2 cannot be equal due to first test,
+    // so if the first one is perceivable, this is bad.
+    if (test(isPerceivableContent, descendants.first().get())) {
+      return false;
+    }
+
+    // Go through descendants until we reach perceivable *content*,
+    // or the second of the nodes
+    descendants = descendants
+      .rest()
+      .skipUntil(or(isPerceivableContent, equals(node1), equals(node2)));
+
+    // if we hit the second node, this is good;
+    // otherwise we've found perceivable content in between.
+    // descendant cannot be empty because it contained at least
+    // node1 and node2 which are different.
+    return test(or(equals(node1), equals(node2)), descendants.first().get());
+  };
+}
+
+/**
+ * @internal
+ *
+ * Find the lowest common ancestor of two nodes:
+ * * get the ancestors chain of both
+ * * go down the chain, from root to nodes, as long as it is the same node
+ * Complexity: linear in the depth of the nodes.
+ */
+export function lowestCommonAncestor(
+  node1: Node,
+  node2: Node,
+  options: Node.Traversal = {}
+): Option<Node> {
+  // Ancestors, from root to nodes.
+  let ancestors1 = node1.inclusiveAncestors(options).reverse();
+  let ancestors2 = node2.inclusiveAncestors(options).reverse();
+
+  // Going down the tree as long as the node is ancestor to both nodes
+  let commonAncestor: Option<Node> = None;
+  let next1 = ancestors1.first();
+  let next2 = ancestors2.first();
+  while (next1.equals(next2)) {
+    commonAncestor = next1;
+    ancestors1 = ancestors1.rest();
+    next1 = ancestors1.first();
+    ancestors2 = ancestors2.rest();
+    next2 = ancestors2.first();
+
+    if (next1.isNone() || next2.isNone()) {
+      // This is needed for the corner case of nodes in different trees but
+      // at the exact same depth…
+      break;
+    }
+  }
+
+  return commonAncestor;
```

Comment:
I've added `Indexed#reduceWhile()` and `Indexed#reduceUntil()` which should help make this bit of code a little more clear.

Reply:
The problem is that we need to move in both lists simultaneously.
If we use a reducer on one, we still need to manually avance through the second and I'm not convinced that breaking the symmetry will help legibility 🤔 

OTOH, the last `if` can be replaced by `next1.isSome()` or `not(ancestors1.isEmpty())` in the `while` condition that may make the intend clearer ("while the lists are not empty and start with the same element"). Also maybe renaming `next1/2` to `first1/2`?

Reply:
I forgot to mention `Indexed#zip()`; first, zip up the ancestor lists and reduce them to the next ancestor while they're equal.